tests/contrib/fastapi/test_fastapi.py

```diff
@@ -0,0 +1,423 @@
+
+    assert response.status_code == 200
+    assert response.json() == {"User Attribute": "Test User"}
+
+    spans = tracer.writer.pop_traces()
+    assert len(spans) == 1
+    assert len(spans[0]) == 1
+    request_span = spans[0][0]
+    assert request_span.service == "fastapi"
+    assert request_span.name == "fastapi.request"
+    assert request_span.resource == "GET /users/{userid:str}/{attribute:str}"
+    assert request_span.error == 0
+    assert request_span.get_tag("http.method") == "GET"
+    assert request_span.get_tag("http.url") == "http://testserver/users/testUserID/name"
+    assert request_span.get_tag("http.status_code") == "200"
+
+
+def test_distributed_tracing(client, tracer):
+    headers = [
+        (http_propagation.HTTP_HEADER_PARENT_ID, "5555"),
+        (http_propagation.HTTP_HEADER_TRACE_ID, "9999"),
+        ("sleep", "False"),
+    ]
+    response = client.get("http://testserver/", headers=dict(headers))
+
+    assert response.status_code == 200
+    assert response.json() == {"Homepage Read": "Success"}
+
+    spans = tracer.writer.pop_traces()
+    assert len(spans) == 1
+    assert len(spans[0]) == 1
+    request_span = spans[0][0]
+    assert request_span.service == "fastapi"
+    assert request_span.name == "fastapi.request"
+    assert request_span.resource == "GET /"
+    assert request_span.error == 0
+    assert request_span.get_tag("http.method") == "GET"
+    assert request_span.get_tag("http.url") == "http://testserver/"
+    assert request_span.get_tag("http.status_code") == "200"
+    assert request_span.parent_id == 5555
+    assert request_span.trace_id == 9999
+
+
+@pytest.mark.asyncio
+async def test_multiple_requests(application, tracer):
+    with override_http_config("fastapi", dict(trace_query_string=True)):
+        async with httpx.AsyncClient(app=application) as client:
+            responses = await asyncio.gather(
+                client.get("http://testserver/", headers={"sleep": "True"}),
+                client.get("http://testserver/", headers={"sleep": "False"}),
+            )
+
+    assert len(responses) == 2
+    assert [r.status_code for r in responses] == [200] * 2
+    assert [r.json() for r in responses] == [{"Homepage Read": "Sleep"}, {"Homepage Read": "Success"}]
+
+    spans = tracer.writer.pop_traces()
+    assert len(spans) == 2
+    assert len(spans[0]) == 1
+    assert len(spans[1]) == 1
+
+    r1_span = spans[0][0]
+    assert r1_span.service == "fastapi"
+    assert r1_span.name == "fastapi.request"
+    assert r1_span.resource == "GET /"
+    assert r1_span.get_tag("http.method") == "GET"
+    assert r1_span.get_tag("http.url") == "http://testserver/"
+
+    r2_span = spans[1][0]
+    assert r2_span.service == "fastapi"
+    assert r2_span.name == "fastapi.request"
+    assert r2_span.resource == "GET /"
+    assert r2_span.get_tag("http.method") == "GET"
+    assert r2_span.get_tag("http.url") == "http://testserver/"
```

Comment:
We probably want to check that the traces aren't the same

```suggestion
    assert r2_span.get_tag("http.url") == "http://testserver/"
    assert r1_span.trace_id != r2_span.trace_id
```